### PR TITLE
Improve performance of time panel

### PR DESCRIPTION
**Pull request by Wumpf**
### Related

* More optimization depends likely on https://github.com/rerun-io/rerun/issues/8221

### What

Make the timepanel faster. Still plenty of room for improvement though.
Achieved by...
* store subscriber to keep track of cumulative chunks needed for a given entity & timeline
    * plus some statistics. less important
* a bit faster `num_events_cumulative_per_unique_time_unsorted`
* improved `DensityGraph::add_range` (this one was surprisingly significant)

### Testing

TL;DR: Minimized timepanel takes less than half the time it use. Other cases are better as well, but more noisy.

----

All testing done on the 2h airtraffic dataset (pre-saved rrd) with all panels hidden and views removed (unless noted otherwise) on my M1 Max. Note that the timeline perf is very dependent on amount of screen real estate - these tests were done maximized on the 14'' Mac's screen.
(Did some throw-away testing on other configs, but these are the ones we're comparing here!)

This round of optimization focused mostly on the "idle" case of having the time panel minimized. There are also some gains for the expanded case, but it's less significant - as illustrated by the profiler screenshots this is largely dominated `num_events_cumulative_per_unique_time` which I hope we can solve with chunk processors (and some untracked overhead I haven't looked into yet).

**Before:**

Frame cpu time without profiler attached hovers around 4.2ms with some outliers.
image

Attaching the profiler doesn't tell us much since the profiler overhead drowns out everything else:
image

Restarting without profiler and expanding the time panel and making it as high as possible gives us about 12ms with frequent spikes beyond 13ms

image

Profiler overhead is ironically not _as_ significant. Averaging a few frames tells us the time panel is at 11.5ms
image

**After**

Frame cpu time without profiler attached hovers between 1.5ms and 2.8ms, rather unsteady
image

Averaging a bunch of frames tells us that the data_density_graph takes now 0.55ms (imho still pretty bad for that it is)
image

Restarting without profiler and expanding the time panel and making it as high as possible gives us around 11ms

Screenshot 2024-11-26 at 15 45 20

(important: this picture hasn't changed otherwise!)

The time panel now takes 9.4ms (that is definitely still very bad!), profiler overhead is still significant but it's manageable:

image

### CODE REVIEW

**Comment by emilk** [comment]
> Attaching the profiler doesn't tell us much since the profiler overhead drowns out everything else

If you are using the latest `puffin_viewer`, you should see a warning when that happens. If you go to the `Table` view you can see which profile scopes are too frequent. Remove them!

**Comment by Wumpf** [comment]
I did, that was in the _before_ state and was actually the first thing I committed :). Didn't bother though to restart my profiling for the PR description at an arbitrary commit and rather used latest main (which moved on quite a bit since I first removed those overhead inducing profile scopes).
I think I no longer got the warning (which I did get before indeed!), but there's still plenty of overhead that I should probably address!

File: crates/store/re_chunk/src/chunk.rs [hunk 1]
```diff
@@ -452,24 +450,24 @@ impl Chunk {
...
+            if let Some(validity) = list_array.validity() {
+                for (time, is_valid) in time_column.times().zip(validity.iter()) {
+                    *result_unordered.entry(time).or_default() += is_valid as u64;
+                }
+            } else {
+                for time in time_column.times() {
+                    *result_unordered.entry(time).or_default() += 1;
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/store/re_chunk/src/chunk.rs:466]: I tried this instead:

```
        let result_unordered = self
            .components
            .values()
            .flat_map(|list_array| {
                if let Some(validity) = list_array.validity() {
                    itertools::Either::Left(time_column.times().zip(validity.iter()))
                } else {
                    itertools::Either::Right(time_column.times().map(|time| (time, true)))
                }
            })
            .fold(HashMap::default(), |mut acc, (time, count)| {
                *acc.entry(time).or_default() += count as u64;
                acc
            });
```

Which caused 2.8ms -> 3.0ms  in a heavy use example (made the time panel large and expanded a few components, causing 62 calls)

[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/store/re_chunk/src/chunk.rs:466]: ah but nested fold is down to 2.7ms. Aka doesn't seem to make a difference
```rust
        let result_unordered =
            self.components
                .values()
                .fold(HashMap::default(), |acc, list_array| {
                    if let Some(validity) = list_array.validity() {
                        time_column.times().zip(validity.iter()).fold(
                            acc,
                            |mut acc, (time, is_valid)| {
                                *acc.entry(time).or_default() += is_valid as u64;
                                acc
                            },
                        )
                    } else {
                        time_column.times().fold(acc, |mut acc, time| {
                            *acc.entry(time).or_default() += 1;
                            acc
                        })
                    }
                });
```
                
guess I use that one then

[the remark this answers is not in the document]
Comment (emilk) [reply crates/store/re_chunk/src/chunk.rs:466]: Maybe construct a proper benchmark for this function?

File: crates/viewer/re_time_panel/src/recursive_chunks_per_timeline_subscriber.rs [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+impl PathRecursiveChunksPerTimelineStoreSubscriber {
+    /// Accesses the global store subscriber.
+    ///
+    /// Lazily registers the subscriber if it hasn't been registered yet.
+    pub fn subscription_handle() -> ChunkStoreSubscriberHandle {
+        static SUBSCRIPTION: OnceCell<ChunkStoreSubscriberHandle> = OnceCell::new();
+        *SUBSCRIPTION.get_or_init(|| ChunkStore::register_subscriber(Box::<Self>::default()))
+    }
+}
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/viewer/re_time_panel/src/recursive_chunks_per_timeline_subscriber.rs:88]: as is this thing is too specific to the timepanel, so I wouldn't want it to live in the entitydb directly. Having a subscription mechanism there instead would be great though

[the remark this answers is not in the document]
Comment (emilk) [reply crates/viewer/re_time_panel/src/recursive_chunks_per_timeline_subscriber.rs:88]: Does late subscribers get all existing chunks?

[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/viewer/re_time_panel/src/recursive_chunks_per_timeline_subscriber.rs:88]: not today, no. Which is why I have to poke the subscriber in `TimePanel::default` to ensure it's up and running before any data comes in

**Pull request merged**